Add support for Marimo notebooks

N283T opened the pull request:
## Description
This PR adds support for running `mols2grid` within [marimo](https://marimo.io) notebooks.

Previously, the grid would not render correctly or support interactivity in marimo due to differences in how JavaScript, HTML, and widgets are handled compared to Jupyter. This implementation fixes the issue by:
1.  Detecting the marimo execution environment.
2.  Forcing the grid to render inside an `iframe` when in marimo. This isolates the grid's styles and scripts, ensuring it displays correctly.
3.  Returning a `marimo.vstack` containing both the `anywidget` instance and the `marimo.Html` grid. This ensures the `anywidget` backend is active, enabling bidirectional communication (e.g., for selection).
4.  Suppressing automatic widget display in `__init__`. In Marimo, we manually return the widget in `display()`, so the automatic display from `__init__` (used in Jupyter) caused double rendering and warnings.

## Changes
- **`mols2grid/utils.py`**: Added `is_running_within_marimo()` utility function.
- **`mols2grid/molgrid.py`**: 
    - Updated `MolGrid.display()` to use `marimo.Html` in an iframe and return it wrapped in `marimo.vstack` with the widget.
    - Updated `MolGrid.__init__` to skip `display(widget)` when running in marimo to prevent double rendering.
- **`tests/test_marimo_integration.py`**: Added unit tests to verify marimo detection and display logic.

## Verification
- Verified that `mols2grid.display(df)` renders the interactive grid correctly in a local marimo notebook.
- Verified that selections made in the grid are correctly reflected in `mols2grid.get_selection()`.
- Added unit tests pass with `uv run pytest`.

N283T committed:
feat: add Marimo integration files and tests, update dependencies, and adjust core utilities.

N283T committed:
feat: Improve Marimo integration by delaying widget display and returning `vstack` with the widget and content.

cbouy commented on mols2grid/molgrid.py:
```diff
@@ -1073,8 +1075,27 @@ def display(
+                    doc=escape(doc),
+                )
+                return mo.vstack([self.widget, mo.Html(iframe)])
+            return mo.vstack([self.widget, mo.Html(doc)])
```
this shouldn't repeat the `use_iframe` block below, instead do somethiing like this to keep it DRY:

```python
requires_marimo = is_running_within_marimo()
if use_iframe or requires_marimo:
    iframe = env.get_template...
    if requires_marimo:
        import marimo as mo

        return mo.vstack...
    return HTML(iframe)
return HTML(doc)
```

cbouy commented on pyproject.toml:
```diff
@@ -28,6 +28,7 @@ classifiers = [
 dependencies = [
     "anywidget",
     "jinja2",
+    "marimo>=0.18.4",
```
marimo shouldn't be part of the base dependencies, there's no reason to require it. It should be added in the tests dependencies instead

cbouy commented:
Thanks @N283T for making a PR, this is greatly appreciated!

Coincidentally, I was planning to add direct marimo support in a refactoring branch #71, but until I manage to get more time to work on this I think we should go ahead with your PR.

I'll enable the CI pipelines, if you could address any issues highlighted there that would be great, and I've already added some minor comments (the main one being that we don't want marimo to be required when installing mols2grid). If you need any help with the above I'm happy to go into more details.

N283T committed:
refactor: make marimo an optional dependency

cbouy commented on pyproject.toml:
```diff
@@ -37,6 +37,9 @@ dynamic = ["version"]
 
+[project.optional-dependencies]
+marimo = ["marimo>=0.18.4"]
+
```
no need to declare it as an optional dependency either, move to `tests` under `[dependency-groups]` below

cbouy reviewed (changes requested):
Sorry I just realised I never actually posted my PR comments, here they are

cbouy commented:
you should be able to fix the ruff error with `uv run poe style-fix`.

N283T committed:
refactor: Consolidate Marimo display logic to use iframes and update dependency configuration.

N283T commented:
Thanks for the review! I've addressed the comments:

- Refactored `MolGrid.display` in `mols2grid/molgrid.py` to avoid code duplication across the `marimo` and `iframe` conditional blocks, following the suggested structure.
- Moved `marimo` from `project.optional-dependencies` to the `tests` dependency group in `pyproject.toml`.
- Ran `uv run poe style-fix` and fixed the corresponding linting errors in the tests.

Additionally, `tests/test_interface.py::test_callbacks_3D` is failing locally for me (missing `<canvas>` element), which seems to be a timing issue unrelated to these changes. Is it okay to proceed with this?

cbouy commented:
Looks good, thanks @N283T ! Some additional small things and then I'll merge:
- Could you create a new `scripts` directory and rename/move the `reproduce_marimo.py` file as `scripts/marimo_example.py` please?
- Could you directly use the `mols2grid.datafiles.SOLUBILITY_SDF` input there instead of a stringIO?
- Finally, to make this example maybe slightly more relevant for marimo, we could mimic what is done in the `filtering` tutorial notebook with sliders, something akin to:
```python
solubility_range = mo.ui.range_slider(
    -10, 2, 0.5, debounce=True, show_value=True, full_width=True, label="Solubility"
)
grid = mols2grid.MolGrid.from_sdf(mols2grid.datafiles.SOLUBILITY_SDF, size=(120, 100))
view = grid.display(n_items_per_page=12)
```
```python
results = grid.dataframe.loc[
    (grid.dataframe["SOL"] >= solubility_range.value[0]) & (grid.dataframe["SOL"] <= solubility_range.value[1])
]
grid.filter_by_index(results.index)
mo.vstack([solubility_range, view])
```

This will make it easier for me to test that all relevant functionalities are preserved when I make changes in the future

N283T committed:
feat: update Marimo examples.

N283T commented:
Thanks! I’ve updated the example following the suggested snippet:

- Moved `reproduce_marimo.py` to `scripts/marimo_example.py`.
- Switched to using `mols2grid.datafiles.SOLUBILITY_SDF`.
- Reworked the example to use a `range_slider` for filtering, matching the filtering tutorial pattern.

Note that grid initialization needs to be kept in a separate cell: in marimo, slider updates trigger re-execution of dependent cells, and recreating the grid on each update would reset the widget state.

N283T commented:
Unrelated follow-up (but still marimo-related):

![mols2grid_get_state](https://github.com/user-attachments/assets/3c04a01a-a9c0-42cd-96e0-b0ea6bb446e6)


I have a local implementation of a small `MolGrid.get_selection_state()` helper
for marimo, which exposes the current grid selection as a reactive state.
This makes it easy to drive components like `mo.ui.table` from the current
selection without adding custom observers in user code.

The helper looks roughly like this:

```python
def get_selection_state(self):
    if not is_running_within_marimo():
        raise RuntimeError("This method is only available in a marimo notebook.")

    import marimo as mo

    get_state, set_state = mo.state([])

    def _on_change(change):
        try:
            sel = ast.literal_eval(change["new"])
            set_state(list(sel.keys()))
        except (ValueError, SyntaxError):
            pass

    if not getattr(self.widget, "_marimo_hooked", False):
        self.widget.observe(_on_change, names=["selection"])
        self.widget._marimo_hooked = True

    return get_state
```

With this in place, an example marimo workflow looks like:

```python
...
# display grid
mo.vstack(
    [solubility_range, view]
)
```

```python
# display table
selected = results[results["mols2grid-id"].isin(sel_ids())]
table = mo.ui.table(
    selected.reset_index(drop=True).drop(columns="img"),
    format_mapping={"mol": mol_to_svg},
    freeze_columns_left=["mols2grid-id", "mol"],
    freeze_columns_right=["SOL"],
    label="Selected Molecules"
)

table
```

<img width="1149" height="868" alt=" 2025-12-18 16 51 02" src="https://github.com/user-attachments/assets/71d7e773-b583-49f9-92fc-9c5234e971f1" />

This helper is not included in the current PR, but if you think it would be
useful, I already have it implemented locally and can either:

- add it to this PR, or
- open a separate follow-up PR

Just let me know which you prefer.

N283T committed:
feat: update Marimo examples.

N283T commented:
Sorry, I accidentally pushed a middle state earlier.
It's been fixed now, and the current commit accurately reflects the intended changes.
This push is unrelated to (`MolGrid.get_selection_state()`).

cbouy commented:
That looks like a great addition, thanks for working on that!
I think it's easier if we do everything in this branch, can you add how to use that in the `marimo_example.py` as well?

N283T committed:
feat: Add `MolGrid.get_selection_state` for Marimo integration and update the Marimo example to display selected molecules.

N283T commented:
## Description

This PR adds first-class support for reactive molecule selection in **marimo** by introducing a small helper on `MolGrid`, and updates the example accordingly.

## Changes

* **Added `MolGrid.get_selection_state()`**
  Introduces a helper method that exposes the current grid selection as a Marimo reactive state.
  This makes it straightforward to drive downstream components (e.g. `mo.ui.table`) from grid selections without manual observer wiring.

* **Updated `scripts/marimo_example.py`**
  The example now demonstrates:

  * filtering with a `range_slider`
  * selecting molecules directly in the grid
  * reactively displaying the selected molecules in a Marimo table using `get_selection_state()`

  To keep the example focused and easy to follow, previously added filters such as `MolWt`, `LogP`, etc. were intentionally removed.

* **Tests**
  Added coverage to ensure `get_selection_state()` behaves correctly inside and outside of a Marimo environment.

## Notes

* The grid creation is intentionally isolated in its own cell in the example.
  In marimo, cells are re-executed whenever their dependencies change, and keeping `MolGrid.from_sdf(...)` separate prevents the widget state from being reset on every slider update.

N283T committed:
refactor: combine patch contexts in Marimo integration test

cbouy commented on mols2grid/molgrid.py:
```diff
@@ -198,7 +199,12 @@ def __init__(  # noqa: PLR0912
         else:
             self._cached_selection = {}
-            register._init_grid(name)
+            if is_running_within_marimo():
```
it would be better to directly skip emitting the warning in `_init_grid` [here](https://github.com/cbouy/mols2grid/blob/master/mols2grid/select.py#L24) instead of catching warnings:
```python
if overwrite and not is_running_within_marimo():
```

cbouy commented on tests/test_marimo_integration.py:
```diff
@@ -63,3 +63,68 @@ def test_display_in_marimo():
+@pytest.mark.usefixtures("mock_marimo_module")
+def test_get_selection_state_inside_marimo():
+    df = pd.DataFrame({"SMILES": ["C"]})
+    mg = MolGrid(df, smiles_col="SMILES")
```
might as well set this up as a fixture at the top of the file since it's used in almost all tests here

cbouy commented on tests/test_marimo_integration.py:
```diff
@@ -63,3 +63,68 @@ def test_display_in_marimo():
+        mg.get_selection_state()
+
+        # Verify observe was called
+        assert mock_observe.called
```
nitpick: `mock_observe.assert_called()`

cbouy commented on mols2grid/molgrid.py:
```diff
@@ -802,6 +808,36 @@ def get_selection(self):
+
+        get_state, set_state = mo.state([])
+
+        def _on_change(change):
+            try:
+                sel = ast.literal_eval(change["new"])
+                set_state(list(sel.keys()))
+            except (ValueError, SyntaxError):
+                pass
```
Just me thinking out loud, definitely NOT something to address in this PR, but in my refactoring branch, 
I might move this to be directly integrated in the `SelectionRegister.selection_updated` in `select.py` with some mechanism to register custom callbacks that would be triggered by `selection_updated`, just so that we don't have multiple observers listening to the same event, e.g.
```python
register.add_callback(lambda sel: set_state(list(sel.keys())))
```

cbouy commented on mols2grid/molgrid.py:
```diff
@@ -802,6 +808,36 @@ def get_selection(self):
             columns=self._extra_columns
         )
 
+    def get_selection_state(self):
```
since this is specific to marimo, it should have a more descriptive name, something like `marimo_selection_getter`?

cbouy commented on scripts/marimo_example.py:
```diff
@@ -7,89 +7,90 @@
+    # Same as:
+    # results = grid.dataframe.loc[
+    #     (grid.dataframe["SOL"] >= solubility_range.value[0]) & (grid.dataframe["SOL"] <= solubility_range.value[1])
+    # ]
```
should be moved one line up as:
```suggestion
    # Same as:
    # grid.dataframe["SOL"] >= solubility_range.value[0]) & (grid.dataframe["SOL"] <= solubility_range.value[1])
```

cbouy commented on scripts/marimo_example.py:
```diff
@@ -7,89 +7,90 @@
-    view = grid.display(n_items_per_page=12)
-    return df, grid, view
+    grid = mols2grid.MolGrid.from_sdf(SOLUBILITY_SDF, size=(120, 100))
+    sel_ids = grid.get_selection_state()
```
nitpick:
```suggestion
    get_selection_ids = grid.get_selection_state()
```

cbouy commented on pyproject.toml:
```diff
@@ -118,6 +118,7 @@ extend-exclude = [
     ".vscode",
     "build",
     "site-packages",
+    "scripts/marimo_example.py",
```
Please revert, I'd rather not ignore formatting/linting without a good reason :sweat_smile:

cbouy reviewed (changes requested):
Really nice addition! Just some minor comments, don't forget to run `uv run poe style-fix`

N283T committed:
refactor: improve Marimo integration by relocating grid overwrite warning logic and renaming selection getter.

N283T commented:
Thanks for the review! Sorry for the bundled reply — I addressed all the comments in one go:

* Moved the warning gating into `SelectionRegister._init_grid` (only warn when `overwrite` and not running in marimo), so we no longer catch warnings in `MolGrid.__init__`.
* Added a shared `MolGrid` fixture in `tests/test_marimo_integration.py` to avoid repeating setup.
* Switched the nitpick to `mock_observe.assert_called()`.
* Renamed the marimo-specific helper to `get_marimo_selection()`.
  I chose this name to stay consistent with the existing `get_selection()` API and the usual Python convention of using verbs for accessors.
  (`marimo_selection_getter` felt a bit more like a variable name to me.)
  Happy to rename it if you prefer `marimo_selection_getter` instead.
* Adjusted the “Same as” comment placement in `scripts/marimo_example.py` per suggestion.
* Reverted the Ruff exclusion for `scripts/marimo_example.py`.

Also ran `uv run poe style-fix`.

cbouy commented:
All good from my side!

Could you just add something in the `docs/contents.md` file to mention the compatibility with Marimo since `v2.2.0` (the next release) with a link to the example in `scripts/`, and also add your name and/or github handle in the Acknowledgement section?

N283T committed:
docs: mention Marimo compatibility and add contributor

N283T commented:
Thanks! I've added a short note about Marimo compatibility in `docs/contents.md` with a link to the example, and added myself to the acknowledgments as well.

cbouy reviewed (approved):
All good, thanks a ton @N283T !
I'm going to make the release today and announce it on socials once the conda build is done (most likely tomorrow), should I add a link to your GitHub account, or is there a Linkedin/X/Bluesky/personal website link that you'd prefer?

N283T commented:
Thanks a lot!
Linking to my GitHub is totally fine, and if possible I’d be happy if you could also include my X account:
https://x.com/tbsng_ktkm0923

Once it’s released, I’ll share it in Japanese and help spread the word in the Japanese community as well !!
